subsys/bluetooth/audio/ascs.c

```diff
@@ -1578,12 +1574,38 @@ int bt_ascs_config_ase(struct bt_conn *conn, struct bt_bap_stream *stream,
 
 	ep = &ase->ep;
 
-	err = ascs_ep_set_codec(ep, codec_cfg->id, sys_le16_to_cpu(codec_cfg->cid),
-				sys_le16_to_cpu(codec_cfg->vid), NULL, 0, &rsp);
-	if (err) {
-		ase_free(ase);
-		return err;
+	if (ep == NULL) {
+		return -EINVAL;
+	}
+
+	struct codec_cap_lookup_id_data lookup_data = {
+		.id = codec_cfg->id,
+		.cid = codec_cfg->cid,
+		.vid = codec_cfg->vid,
+	};
+
+	bt_pacs_cap_foreach(ep->dir, codec_lookup_id, &lookup_data);
+
+	if (lookup_data.codec_cap == NULL) {
+		LOG_DBG("Codec with id %u for dir %s is not supported by our capabilities",
+			codec_cfg->id, bt_audio_dir_str(ep->dir));
+		return -ENOENT;
+	}
+
+	(void)memcpy(&ep->codec_cfg, codec_cfg, sizeof(ep->codec_cfg));
+
+#if CONFIG_BT_AUDIO_CODEC_CFG_MAX_DATA_COUNT > 0 && CONFIG_BT_AUDIO_CODEC_MAX_DATA_LEN > 0
+	/* Need to update the `bt_data.data` pointer to the new value after copying the codec */
+	for (size_t i = 0U; i < ARRAY_SIZE(ep->codec_cfg.data); i++) {
+		const struct bt_audio_codec_data *codec_data = &codec_cfg->data[i];
+		struct bt_audio_codec_data *data = &ep->codec_cfg.data[i];
+		const uint8_t data_len = codec_data->data.data_len;
+
+		data->data.data = data->value;
+		data->data.data_len = data_len;
+		memcpy(data->value, codec_data->data.data, data_len);
 	}
+#endif /* CONFIG_BT_AUDIO_CODEC_CFG_MAX_DATA_COUNT > 0 && CONFIG_BT_AUDIO_CODEC_MAX_DATA_LEN > 0 */
```

Comment:
I suppose we will need to copy the metadata as well, right? (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/audio/shell/audio.h#L288-L300)

Reply:
I would probably suggest to move this copy into a function to increase readability, and make easier to remove once https://github.com/zephyrproject-rtos/zephyr/pull/58331 gets merged (at which point it can be removed again).